File: apps/android-usage/companion/src/main/kotlin/com/healthmes/companion/notify/NotificationGrammar.kt
package com.healthmes.companion.notify

import com.healthmes.api.AlertItem
import com.healthmes.briefing.BriefingDisplayState
import com.healthmes.briefing.GlanceBriefing

/**
 * The docs/PLAN.md §8.5 notification grammar, as data:
 *
 * ```
 * [observation, 1 line]
 * [evidence, 1 line]
 * [proposal, 1 line]
 * [buttons]  Apply / Adjust / Keep as is   (real schedule-proposal actions)
 * [link]     Why this? -> decision viewer URL
 * ```
 *
 * Pure Kotlin so the mapping from server payloads is JVM unit-testable.
 *
 * Two sources, in preference order:
 * - [fromAlert] — a `GET /v1/alerts` item carrying the REAL grammar lines the
 *   trigger recorded at fire time (observation `summary`, `evidence` facts,
 *   `proposal`).
 * - [fromGlance] — fallback from the glance payload alone (only the
 *   observation is real there; evidence/proposal are honest fillers proving
 *   the rendering path). Wording rules stay the healthcare domain expert's
 *   deliverable (docs/design/WATCH-NOTIFICATIONS.ko.md).
 */
data class NotificationGrammar(
    val alertId: String?,
    val alertRuleId: String,
    val observation: String,
    val evidence: String,
    val proposal: String,
    val decisionUrl: String?,
    val proposalId: String?,
    val alertRevision: String?,
) {

    /** Three grammar lines for BigTextStyle. */
    fun bigText(): String = "$observation\n$evidence\n$proposal"

    companion object {

        /** Real grammar lines from an alert-history item (issue #10). */
        fun fromAlert(alert: AlertItem): NotificationGrammar = NotificationGrammar(
            alertId = alert.id,
            alertRuleId = alert.ruleId,
            observation = alert.summary,
            evidence = alert.evidenceLine() ?: "Rule ${alert.ruleId} fired",
            proposal = alert.proposal
                ?: "Open the decision record for the reasoning, or adjust in the app.",
            decisionUrl = alert.decisionUrl,
            proposalId = alert.proposalId,
            alertRevision = "${alert.id}:${alert.proposalId ?: "informational"}",
        )

        /** Null when the payload has no top alert to phrase. */
        fun fromGlance(briefing: GlanceBriefing): NotificationGrammar? {
            val top = briefing.alerts.top ?: return null
            val score = briefing.energy.score?.toString() ?: BriefingDisplayState.NO_SCORE
            val evidence =
                "Rule ${top.ruleId} fired · ${briefing.alerts.unresolvedCount} unresolved " +
                    "in 24h · energy $score (${briefing.energy.confidence})"
            val proposal =
                if (top.decisionUrl != null) {
                    "Open the decision record for the reasoning; reply in Telegram to adjust."
                } else {
                    "Reply in Telegram to review and adjust today's plan."
                }
            return NotificationGrammar(
                alertId = top.id,
                alertRuleId = top.ruleId,
                observation = top.summary,
                evidence = evidence,
                proposal = proposal,
                decisionUrl = top.decisionUrl,
                proposalId = null,
                alertRevision = top.id?.let { "$it:informational" },
            )
        }

        fun pendingNotification(
            previousRevisions: Set<String>,
            alerts: List<NotificationGrammar>,
        ): NotificationGrammar? = alerts.firstOrNull { alert ->
            val id = alert.alertId
            val revision = alert.alertRevision
            if (id == null || revision == null || revision in previousRevisions) {
                false
            } else {
                val hasPriorRevision = previousRevisions.any { it.startsWith("$id:") }
                !hasPriorRevision || alert.proposalId != null
            }
        }

        fun legacyRevisions(alerts: List<NotificationGrammar>): Set<String> =
            alerts.mapNotNull { alert ->
                alert.alertId?.let { "$it:informational" }
            }.toSet()

        fun currentRevisions(alerts: List<NotificationGrammar>): Set<String> =
            alerts.mapNotNull { it.alertRevision }.toSet()

        fun mergeRevisions(
            previousRevisions: Set<String>,
            alerts: List<NotificationGrammar>,
            notifiedRevision: String?,
        ): Set<String> {
            val currentIds = alerts.mapNotNull { it.alertId }.toSet()
            return buildSet {
                addAll(previousRevisions.filter { it.substringBefore(':') in currentIds })
                for (alert in alerts) {
                    val revision = alert.alertRevision ?: continue
                    if (
                        revision in previousRevisions ||
                        revision == notifiedRevision
                    ) {
                        add(revision)
                    }
                }
            }
        }
    }
}

File: apps/windows-companion/src/HealthMes.Glance.Core/NotificationGrammar.cs
using System.Text.Json;

namespace HealthMes.Glance.Core;

/// <summary>
/// The docs/PLAN.md §8.5 notification grammar, as data:
///
/// <code>
/// [observation, 1 line]
/// [evidence, 1 line]
/// [proposal, 1 line]
/// [link]     Why this? -> decision viewer URL
/// </code>
///
/// Pure C# so the mapping from a payload is unit-testable on any OS. The
/// glance mapping mirrors the Android companion's NotificationGrammar.kt
/// verbatim (same evidence/proposal wording) so the two desktop/mobile
/// surfaces phrase the same alert the same way.
///
/// PLACEHOLDER CONTENT: wording rules are the healthcare domain expert's
/// deliverable (docs/design/WATCH-NOTIFICATIONS.ko.md). Buttons (✅✏️❌) are
/// a phone/Telegram surface — the desktop toast carries the observation/
/// evidence/proposal lines plus the "why this?" decision link, which §1.1 of
/// the worksheet says no surface may drop.
/// </summary>
public sealed record NotificationGrammar
{
    public required string Observation { get; init; }

    public required string Evidence { get; init; }

    public required string Proposal { get; init; }

    public required string? DecisionUrl { get; init; }

    /// <summary>The three grammar lines, one per line (toast/flyout body).</summary>
    public string BodyText() => $"{Observation}\n{Evidence}\n{Proposal}";

    /// <summary>Null when the payload has no top alert to phrase.</summary>
    public static NotificationGrammar? FromGlance(GlancePayload payload)
    {
        var top = payload.Alerts.Top;
        if (top is null)
        {
            return null;
        }
        var score = GlanceFormat.ScoreText(payload.Energy.Score);
        var evidence =
            $"Rule {top.RuleId} fired · {payload.Alerts.UnresolvedCount} unresolved in 24h" +
            $" · energy {score} ({payload.Energy.Confidence.WireName()})";
        var proposal = top.DecisionUrl is not null
            ? "Open the decision record for the reasoning; reply in Telegram to adjust."
            : "Reply in Telegram to review and adjust today's plan.";
        return new NotificationGrammar
        {
            Observation = top.Summary,
            Evidence = evidence,
            Proposal = proposal,
            DecisionUrl = top.DecisionUrl,
        };
    }

    /// <summary>
    /// Grammar for one `GET /v1/alerts` item. Unlike the glance top alert,
    /// alert-history items carry the REAL evidence facts and proposal line
    /// recorded at fire time — those are used verbatim when present.
    /// </summary>
    public static NotificationGrammar FromAlert(AlertItem alert)
    {
        var evidence = RenderEvidence(alert.Evidence) ?? $"Rule {alert.RuleId} fired";
        var proposal = alert.Proposal
            ?? (alert.DecisionUrl is not null
                ? "Open the decision record for the reasoning; reply in Telegram to adjust."
                : "Reply in Telegram to review and adjust today's plan.");
        return new NotificationGrammar
        {
            Observation = alert.Summary,
            Evidence = evidence,
            Proposal = proposal,
            DecisionUrl = alert.DecisionUrl,
        };
    }

    /// <summary>
    /// "hrv_delta_pct -18 · baseline_days 14" out of the evidence facts
    /// object; null when there is nothing renderable.
    /// </summary>
    public static string? RenderEvidence(JsonElement? evidence)
    {
        if (evidence is not { ValueKind: JsonValueKind.Object } facts)
        {
            return null;
        }
        var parts = new List<string>();
        foreach (var property in facts.EnumerateObject())
        {
            var value = property.Value.ValueKind switch
            {
                JsonValueKind.String => property.Value.GetString(),
                JsonValueKind.Number or JsonValueKind.True or JsonValueKind.False =>
                    property.Value.GetRawText(),
                _ => null, // nested objects/arrays/nulls stay off the one-line surface
            };
            if (value is not null)
            {
                parts.Add($"{property.Name} {value}");
            }
        }
        return parts.Count > 0 ? string.Join(" · ", parts) : null;
    }
}
